text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,1563 @@
+   previously needed to. This especially holds in pattern contexts.
+   This will be made clear later on in this RFC.
+
+5. Turbofish is not always possible! Consider for example:
+
+   ```rust
+   fn display_all(elems: impl Iterator<Item: Display>) { ... }
+   ```
+
+   As of today (2018-08-06), it is not possible to use turbofish at all
+   as you could have done had you instead written:
+
+   ```rust
+   fn display_all<I: Iterator<Item: Display>>(elems: I) { ... }
+   ```
+
+   While this may change in the future, it may also be the case that anonymous
+   `arg: impl Trait`s can't ever be turbofished. In such a case, type ascription
+   is our saving grace; it works independently of how `display_all` is defined
+   and let's us easily constrain the type of `elems` in a syntactically
+   light-weight way in any case.
+
+6. Type ascription is helpful when doing *type* driven development
+   and opens up more possibilities to move in the direction of
+   interactive development as is possible with [agda-mode] and [idris-mode].
+
+[agda-mode]: http://agda.readthedocs.io/en/v2.5.2/tools/emacs-mode.html
+[idris-mode]: https://github.com/idris-hackers/idris-mode
+
+## Type ascription has already been accepted as an RFC
+
+We noted previously that [RFC 803] already accepted type ascription in
+expression contexts. Thus, we have already collectively deemed to some extent
+that ascription is something we want. However, the previous RFC did not apply
+ascription uniformly. We believe it is beneficial to do so. We also believe
+that much of the motivation for accepting RFC 803 applies to the extensions
+proposed here.
+
+## More DRY code
+
+By introducing type ascription as a pattern, we can simplify function definitions
+to permit the following:
+
+```rust
+fn take_wrapping(Wrapping(count: usize)) -> R { .. }
+```
+
+instead of the following:
+
+```rust
+fn take_wrapping(Wrapping(count: usize): Wrapping<usize>) -> R { .. }
+```
+
+Compared to type ascription in expression contexts or using `let` bindings
+we can also get away with leaving more inferred when pattern matching.
+For example, consider:
+
+```rust
+let temporary: Option<Vec<u8>> = expr;
+match temporary {
+    None => logic,
+    Some(vec) => logic,
+}
+```
+
+as compared to:
+
+```rust
+match expr : Option<Vec<u8>> {
+    None => logic,
+    Some(vec) => logic,
+}
+```
+
+and against:
+
+```rust
+match expr {
+    None => logic,
+    Some(vec: Vec<u8>) => logic,
+}
+```
+
+In the last case, the typing annotation is both *most local* and also does not
+require you to annotate information that is both obvious to the reader
+(who is familiar with `Option<T>`) and to the compiler
+(that `expr : Option<?T>` for some `?T`).
+Because the annotation is more local, we can employ more local reasoning.
+This is particularly useful if the `enum` contains many variants in which
+case the type ascription on `expr` may not be immediately visible.
```

Comment:
I really like this example, and would like to see the multi-variant thing more.  Maybe something like

```rust
match foo.parse() {
    Ok(x: i32) => ...,
    Err(e) => ...
}
```

as opposed to `::<i32>` or `: Result<i32, _>`.

Reply:
This is a great example; I'll be sure to include it.

Reply:
Thanks!  Update looks great.